### ISSUE TITLE
This condition will always return 'false' since the types '"string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"' and '"array"' have no overlap?

### ISSUE THREAD

**Issue by ghost** [microsoft/TypeScript#40196]
TypeScript is complaining about this code:

``` ts
function isThisJsonAnObjectOrArray(json: string) {
    let data = JSON.parse(json);
    return (typeof data === 'object' || typeof data === 'array');
}
```

With an error:

```
This condition will always return 'false' since the types '"string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"' and '"array"' have no overlap. (2367)
```

But how would it always return false? I can see how it would be false if `data` is a string for example, but if it's an object or array why would it not return true?

**Comment by AlCalzone** [comment]
You'll probably want to use `Array.isArray(data)` or if that doesn't exist in your runtime `Object.prototype.toString.call(data) === "[object Array]"`
